src/ops/indexing.rs

```diff
@@ -0,0 +1,1799 @@
+        if item.is_array_or_index() {
+            if have_array_or_index && have_non_array {
+                gather_first = true;
+                break;
+            }
+            have_array_or_index = true;
+        } else {
+            have_non_array = have_non_array || have_array_or_index;
+        }
+    }
+
+    let array_count = operations.iter().filter(|op| op.is_array()).count();
+    let have_array = array_count > 0;
+
+    let mut remaining_indices: Vec<ArrayIndexOp> = Vec::new();
+    if have_array {
+        // apply all the operations (except for .newAxis) up to and including the
+        // final .array operation (array operations are implemented via gather)
+        let last_array_or_index = operations
+            .iter()
+            .rposition(|op| op.is_array_or_index())
+            .unwrap(); // safe because we know there is at least one array operation
+                       // let remaining = indices.split_off(last_array_or_index + 1);
+        let gather_indices = operations[..=last_array_or_index]
+            .iter()
+            .filter(|op| !matches!(op, Ellipsis | ExpandDims));
+        let (max_dims, gathered) = gather_nd(&src, gather_indices, gather_first, stream.clone());
+
+        src = Cow::Owned(gathered);
+
+        // Reassemble the indices for the slicing or reshaping if there are any
+        if gather_first {
+            remaining_indices.extend((0..max_dims).map(|_| ArrayIndexOp::range_full()));
+
+            // copy any newAxis in the gatherIndices through.  any slices get
+            // copied in as full range (already applied)
+            for item in &operations[..=last_array_or_index] {
+                match item {
+                    ExpandDims => remaining_indices.push(item.clone()),
+                    Slice { .. } => remaining_indices.push(ArrayIndexOp::range_full()),
+                    _ => {}
+                }
+            }
+
+            // append the remaining operations
+            remaining_indices.extend(operations[(last_array_or_index + 1)..].iter().cloned());
+        } else {
+            // !gather_first
+            for item in operations.iter() {
+                match item {
+                    TakeIndex { .. } | TakeArray { .. } => break,
+                    ExpandDims => remaining_indices.push(item.clone()),
+                    _ => remaining_indices.push(ArrayIndexOp::range_full()),
+                }
+            }
+
+            // handle the trailing gathers
+            remaining_indices.extend((0..max_dims).map(|_| ArrayIndexOp::range_full()));
+
+            // and the remaining operations
+            remaining_indices.extend(operations[(last_array_or_index + 1)..].iter().cloned());
+        }
+    }
+
+    if have_array && remaining_indices.is_empty() {
+        // `clone` returns a new array with the same shape and data
+        match src {
+            Cow::Borrowed(src) => return src.clone(),
+            Cow::Owned(src) => return src,
+        }
```

Comment:
no biggie - small suggestion

```suggestion
        return match src {
            Cow::Borrowed(src) => src.clone(),
            Cow::Owned(src) => src,
        }
```

Reply:
Addressed in 55ca17d931a3dff70b189ecb3086957543ca4d55